test/integration_test/s3/getObject.spec.js

```diff
@@ -0,0 +1,209 @@
+    }).then(function() {
+      return s3.waitFor('objectNotExists', {
+        Bucket: params.Bucket,
+        Key: params.Key,
+      }).promise()
+    }).then(function() {
+      return s3.deleteBucket({Bucket: params.Bucket}).promise();
+    }).then(function(data) {
+      return s3.waitFor('bucketNotExists', {Bucket: params.Bucket}).promise();
+    }).then(function() {
+      if (typeof done === 'function') done();
+    }).catch(function(err) {
+      throw new Error('Cannot delete bucket: ' + err);
+      exit(1);
+    })
+  }
+
+  before('setup s3 client with responseChecksumAlgorithm equals \'md5\'', function() {
+    s3 = new AWS.S3({responseChecksumAlgorithm: 'md5'});
+  })
+
+  beforeEach('setup bucket and object...', function(done) {
+    params = {
+      Bucket: bucketName,
+      Key: 'key',
+    };
+    putObject({Body: 'this is a test!'}, done);
+  })
+
+  afterEach('delete bucket...', function(done) {
+    this.timeout(30000);
+    delectBucket(done);
+  })
+
+  it('should correctly call callback when success', function(done) {
+    s3.getObject({
+      Bucket: params.Bucket,
+      Key: params.Key,
+      Range: 'bytes=0-5',
+    }, function(err, data) {
+      expect(err).to.be['null'];
+      expect(data).to.not.be['null'];
+      expect(data.Body.toString()).to.equal('this i');
+      expect(data.ContentLength).to.equal(6);
+      expect(data.Range).to.equal('bytes 0-5/15')
+      done();
+    });
+  });
+
+  it('should correctly call callback when fail', function(done) {
+    s3.getObject({
+      Bucket: params.Bucket,
+      Key: 'wrongkey',
+    }, function(err, data) {
+      expect(data).to.be['null'];
+      expect(err).to.not.be['null'];
+      expect(err.code).to.equal('NoSuchKey');
+      done();
+    });
+  });
+
+  it('should correctly create read stream', function(done) {
+    var data = '';
+    var error = null;
+    var s = s3.getObject({
+      Bucket: params.Bucket,
+      Key: params.Key,
+    }).createReadStream();
+    s.on('readable', function() {
+      var dat = s.read();
+      if(dat) {
+        data += dat;
+      }
+    })
+    s.on('error', function(err) {
+      error = err;
+    })
+    s.on('end', function() {
+      expect(error).to.be['null'];
+      expect(data).to.not.be['null'];
+      expect(data).to.equal('this is a test!');
+      done();
+    })
+  });
+
+  it('should raise error in creating read stream when fail', function(done) {
+    var data = '';
+    var error = null;
+    var s = s3.getObject({
+      Bucket: params.Bucket,
+      Key: 'wrongKey',
+    }).createReadStream();
+    s.on('readable', function() {
+      var dat = s.read();
+      if(dat) {
+        data += dat;
+      }
+    })
+    s.on('error', function(err) {
+      error = err;
+      expect(error).to.not.be['null'];
+      expect(error.code).to.equal('NoSuchKey');
+      done();
+    })
+    s.on('end', function() {
+      expect(data).to.equal('');
```

Comment:
Does `end` ever get triggered? If it shouldn't, maybe we should throw an error here. If it should, then `done()` should land here assuming `end` is called after `error`.